src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,362 @@
+                {
+                    var newFuncs = functionSet.WithInvariantNameInternal(key);
+                    var fInvariantList = WithInvariantNameInternal(key);
+
+                    if (fInvariantList.Any())
+                    {
+                        fInvariantList.AddRange(newFuncs);
+                    }
+                    else
+                    {
+                        _functionsInvariant.Add(key, newFuncs);
+                    }
+                }
+
+                foreach (var key in functionSet.Namespaces)
+                {
+                    var newFuncs = functionSet.WithNamespaceInternal(key);
+                    var fnsList = WithNamespaceInternal(key);
+
+                    if (fnsList.Any())
+                    {
+                        fnsList.AddRange(newFuncs);
+                    }
+                    else
+                    {
+                        _namespaces.Add(key, newFuncs);
+                    }
+                }
+
+                _enums.AddRange(functionSet._enums);
+            }
+
+            return this;
+        }
+
+        internal void Add(IEnumerable<TexlFunction> functions)
+        {
+            if (functions == null)
+            {
+                throw new ArgumentNullException(nameof(functions));
+            }
+
+            foreach (var t in functions)
+            {
+                Add(t);
+            }
+        }
+
+        internal int Count()
+        {
+            return _count;
+        }
+
+        internal bool Any()
+        {
+            return _count != 0;
+        }
+
+        internal bool AnyWithName(string name) => _functions.ContainsKey(name);
+
+        internal void RemoveAll(string name)
+        {
+            if (_functions.TryGetValue(name, out List<TexlFunction> removed))
+            {
+                _count -= removed.Count();
+                _functions.Remove(name);
+            }
+
+            if (_functionsInvariant.TryGetValue(name, out _))
+            {
+                _functionsInvariant.Remove(name);
+            }
+
+            if (removed != null && removed.Any())
+            {
+                foreach (TexlFunction f in removed)
+                {
+                    List<TexlFunction> fnsList = WithNamespaceInternal(f.Namespace);
+
+                    if (fnsList.Count == 1)
+                    {
+                        _namespaces.Remove(f.Namespace);
+                    }
+                    else
+                    {
+                        fnsList.Remove(f);
+                    }
+
+                    f.GetRequiredEnumNames().ToList().ForEach(removedEnum => _enums.Remove(removedEnum));
+                }
+            }
+
+            // If functions with the given name aren't found, this is ok.
+        }
+
+        internal void RemoveAll(TexlFunction function)
+        {
+            if (_functions.TryGetValue(function.Name, out List<TexlFunction> funcs))
```

Comment:
>List<TexlFunction> [](http://example.com/codeflow?start=58&length=18)

I'd suggest using `var` more for `out` params, the type `List<TexlFunction> is repeated a bunch here and seems pretty clear to me. 

But, that's a personal preference, feel free to ignore.

Reply:
I personally hate 'var' while I admit this is convenient at times.
Here, I would say specifying the type is 'mandatory' as it makes the code much clearer. When you look at the code, if you don't remember the exact type for _functions, you can't know the type of funcs. With the 's' at the end, you could imagine it's an IEnumerable can can't know if that's a List, Array, Dictionary and what objects are stored there.
Now, I agree this is repetitive with lines 331 and 341 but I find it strange if I'd have used one form for line 320 and another for the rest (= makes it suspicious!)
As you said, it's a personal perference...